packages/react-reconciler/src/ReactFiberNewContext.new.js

```diff
@@ -311,6 +345,214 @@ export function propagateContextChange<T>(
+                  // the future, we could add back a dirty flag as an
+                  // optimization to avoid double checking, but until we have
+                  // selectors it's not really worth the trouble.
+                  consumer.lanes = mergeLanes(consumer.lanes, renderLanes);
+                  const alternate = consumer.alternate;
+                  if (alternate !== null) {
+                    alternate.lanes = mergeLanes(alternate.lanes, renderLanes);
+                  }
+                  scheduleWorkOnParentPath(consumer.return, renderLanes);
+
+                  // Since we already found a match, we can stop traversing the
+                  // dependency list.
+                  break findChangedDep;
+                }
+              }
+              dep = dependency.next;
+            }
+          }
+        }
+
+        if (nextFiber === workInProgress) {
+          // We're back to the root of this subtree. Exit.
+          nextFiber = null;
+          break;
+        }
+
+        //Traverse to next sibling.
+        const sibling = nextFiber.sibling;
+        if (sibling !== null) {
+          // Set the return pointer of the sibling to the work-in-progress fiber.
+          sibling.return = nextFiber.return;
+          nextFiber = sibling;
+          break;
+        }
+        // No more siblings. Traverse up.
+        nextFiber = nextFiber.return;
+      }
+    }
+    fiber = nextFiber;
+  }
+}
+
+// Alias for propagating a deferred tree (Suspense, Offscreen). Currently it's
+// the same algorithm but there may be a way to optimize one or the other.
+export const propagateParentContextChangesToDeferredTree = lazilyPropagateParentContextChanges;
+
+export function lazilyPropagateParentContextChanges(
+  current: Fiber,
+  workInProgress: Fiber,
+  renderLanes: Lanes,
+) {
+  if (!enableLazyContextPropagation) {
+    return false;
+  }
+
+  // Collect all the parent providers that changed. Since this is usually small
+  // number, we use an Array instead of Set.
+  let contexts = null;
+  let parent = workInProgress;
+  while (parent !== null && (parent.flags & DidPropagateContext) === NoFlags) {
+    if (parent.tag === ContextProvider) {
+      const currentParent = parent.alternate;
+      if (currentParent !== null) {
+        const oldProps = currentParent.memoizedProps;
+        if (oldProps !== null) {
+          const providerType: ReactProviderType<any> = parent.type;
+          const context: ReactContext<any> = providerType._context;
+
+          const newProps = parent.pendingProps;
+          const newValue = newProps.value;
+
+          const oldValue = oldProps.value;
+
+          const changedBits = calculateChangedBits(context, newValue, oldValue);
+          if (changedBits !== 0) {
+            if (contexts !== null) {
+              contexts.push(context, changedBits);
+            } else {
+              contexts = [context, changedBits];
+            }
```

Comment:
If you included contexts that had 0 changedBits you could avoid the mask-breaking problem where a higher provider with changed bits should be masked by a nearer provider without changed bits. the dep would hit the nearest one first and see changedBits of 0 and never get to the higher provider instance

downside is we'd have all contexts since last propagation instead of just the changing ones